ops/src/purge/errored_txs_purger.rs

```diff
@@ -0,0 +1,296 @@
+            &[b"asset", &self.tree, &leaf_idx.to_le_bytes()],
+            &mpl_bubblegum::ID,
+        );
+        let asset_bytes = asset.to_bytes().to_vec();
+        let multi_txn = conn.begin().await?;
+
+        asset_data::Entity::delete_by_id(asset_bytes.clone())
+            .exec(&multi_txn)
+            .await?;
+
+        asset::Entity::delete_by_id(asset_bytes.clone())
+            .exec(&multi_txn)
+            .await?;
+
+        asset_creators::Entity::delete_many()
+            .filter(asset_creators::Column::AssetId.eq(asset_bytes.clone()))
+            .exec(&multi_txn)
+            .await?;
+
+        asset_authority::Entity::delete_many()
+            .filter(asset_authority::Column::AssetId.eq(asset_bytes.clone()))
+            .exec(&multi_txn)
+            .await?;
+
+        asset_grouping::Entity::delete_many()
+            .filter(asset_grouping::Column::AssetId.eq(asset_bytes))
+            .exec(&multi_txn)
+            .await?;
+
+        // Remove all txs for the asset from cl_audits_v2 so that the asset can be re-indexed by the backfiller
+        cl_audits_v2::Entity::delete_many()
+            .filter(cl_audits_v2::Column::LeafIdx.eq(leaf_idx))
+            .filter(cl_audits_v2::Column::Tree.eq(self.tree.clone()))
+            .exec(&multi_txn)
+            .await?;
+
+        multi_txn.commit().await?;
+
+        Ok(())
+    }
+}
+
+#[derive(Debug, Parser, Clone)]
+pub struct PurgeErrTxsArgs {
+    /// The list of trees to verify. If not specified, all trees will be crawled.
+    #[arg(long, env, use_value_delimiter = true)]
+    pub only_trees: Option<Vec<String>>,
+    #[clap(flatten)]
+    pub purge_args: PurgeArgs,
+}
+
+pub async fn start_errored_txs_purge(args: PurgeErrTxsArgs, db: PgPool, rpc: Rpc) -> Result<()> {
+    let start = tokio::time::Instant::now();
+
+    let context = BubblegumContext::new(db.clone(), rpc.clone());
+    let trees = if let Some(ref only_trees) = args.only_trees {
+        TreeResponse::find(&context.solana_rpc, only_trees.clone()).await?
+    } else {
+        TreeResponse::all(&context.solana_rpc).await?
+    };
+
+    let purge_worker_count = args.purge_args.purge_worker_count as usize;
+
+    for tree in trees {
+        let tree_bytes = tree.pubkey.as_ref().to_vec();
+        let (paginate_sender, paginate_receiver) = unbounded_channel::<Vec<TxQueryResult>>();
+        let (mark_sender, mut mark_receiver) = unbounded_channel::<i64>();
+
+        let paginate_handle = Paginate::<PgPool>::build()
+            .pool(db.clone())
+            .batch_size(args.purge_args.batch_size)
+            .sender(paginate_sender)
+            .tree(tree_bytes.clone())
+            .start();
+
+        let mark_handle = MarkDeletion::build()
+            .rpc(rpc.clone())
+            .receiver(paginate_receiver)
+            .sender(mark_sender)
+            .concurrency(args.purge_args.mark_deletion_worker_count as usize)
+            .start();
+
+        let db = db.clone();
```

Comment:
This is a spot where we can add concurrency. Can we structure this where the pagination are ran in threads over the trees.

Actually looking more at we don't really need to check RPC for the trees. We if we don't have the tree than we don't have a failed transaction. This version of paginate should take a list of trees and change the pagination to be a `WHERE IN ()` this lets us flatten out the purge workers to single pagination, mark deletion, and purge threads.